eos-payg-ctl/eos-payg-ctl

```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/python3
+import argparse
+import contextlib
+import datetime as dt
+import sys
+from dateutil import tz
+from gi.repository import GLib, Gio
+
+BUS_NAME = 'com.endlessm.Payg1'
+OBJECT_PATH = '/com/endlessm/Payg1'
+INTERFACE = 'com.endlessm.Payg1'
+ERROR_DOMAIN = 'com.endlessm.Payg1.Error'
+
+DBUS_PROPERTIES_INTERFACE = 'org.freedesktop.DBus.Properties'
+
+
+def __format_utc_timestamp(timestamp):
+    expiry_utc = dt.datetime.fromtimestamp(timestamp, tz=tz.tzutc())
+    expiry_local = expiry_utc.astimezone(tz.tzlocal())
+    return str(expiry_local)
+
+
+def __get_proxy():
+    return Gio.DBusProxy.new_for_bus_sync(
+        Gio.BusType.SYSTEM,
+        Gio.DBusProxyFlags.NONE,
+        None,
+        BUS_NAME,
+        OBJECT_PATH,
+        INTERFACE,
+        None,  # cancellable
+    )
+
+
+def command_status(proxy):
+    props = {
+        name: proxy.get_cached_property(name).unpack()
+        for name in proxy.get_cached_property_names()
+    }
+
+    width = max(map(len, props))
+    formatters = {
+        'ExpiryTime': __format_utc_timestamp,
+        'RateLimitEndTime': __format_utc_timestamp,
+    }
+
+    for key, value in sorted(props.items()):
+        try:
+            formatter = formatters[key]
+        except KeyError:
+            formatted = value
+        else:
+            formatted = formatter(value)
+        print('{:>{}}: {}'.format(key, width, formatted))
+
+
+@contextlib.contextmanager
+def __exit_on_payg_error():
+    try:
+        yield
+    except GLib.GError as e:
+        # The Right Thing To Do would be to make libeos-payg introspectable, so
+        # we can load it and get its error domain & codes registered with
+        # GDBus. This seems like a lot of work for minimal gain, and
+        # hard-coding them here just guarantees they will drift out of sync.
+        remote_error = Gio.dbus_error_get_remote_error(e)
+        if not remote_error or not remote_error.startswith(ERROR_DOMAIN):
+            raise
+
+        # Gio.dbus_error_get_remote_error(e) does not modify e.message,
+        # presumably because the Python-side GLib.GError is marshalled back to
+        # a new temporary C-side GError.
+        message = e.message[e.message.index(remote_error):]
+        print(message, file=sys.stderr, flush=True)
+        raise SystemExit(1)
+
+
+def command_add_code(proxy, code):
+    with __exit_on_payg_error():
+        proxy.AddCode('(sss)', code, code, code)
+
+
+def command_clear_code(proxy):
+    with __exit_on_payg_error():
+        proxy.ClearCode()
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.set_defaults(function=command_status)
+    subparsers = parser.add_subparsers()
+
+    status = subparsers.add_parser('status')
+    status.set_defaults(function=command_status)
+
+    add_code = subparsers.add_parser('add-code')
+    add_code.set_defaults(function=command_add_code)
+    add_code.add_argument('code')
```

Comment:
Could do with a description. The rest of the arguments could do with some descriptions and help too.